sendreceive.go

```diff
@@ -57,8 +61,12 @@ func (client *Client) send(cmd CIPCommand, msgs ...any) error {
 		}
 		n, err := client.conn.Write(b[written:])
 		if err != nil {
-			err2 := client.disconnect()
-			return fmt.Errorf("%w: %v", err, err2)
+			err = fmt.Errorf("problem writing to socket: %w", err)
+			err2 := client.Disconnect()

```

Comment:
This call to disconnect (and any others in this function) create a lock issue - the mutex on the lock is already held when we get here, so Disconnect() cannot get it to send the disconnect message.  I think disconnect used to be split into two parts one of which was just the "housekeeping" disconnect and one which actually send the "graceful" disconnect request to the controller.

Reply:
Good catch on this. I also realized that I messed up when entering my test code command so no wonder why I thought all the tests pass when I hadn't recursively scanned folders for tests. Just so you know, as part of this I realized that some of the test parameters needed to be updated (which will also come in the next string of updates) so that way the tests could be run on not your controller more easily i.e. expanding the test_config.json file.

I am working on a fix for this that separates out some of the functionality for send_and_recv so that way there isn't a dead goroutine. It is being a little tricky so it will probably be sometime tomorrow when I finish and upload all the mods.

As a side note, the graceful disconnect is still attempted in the Disconnect method. I consolidated the disconnect and the Disconnect methods as when I started combining things the separate methods just didn't really make sense to me anymore.

Reply:
Sounds good on all counts.

Reply:
Would you mind checking to make sure that your AB code is up to date? I haven't fully debugged it, but it felt like some of the tags were missing (the one I was specifically looking for was called testdintarr I think but I didn't see it in the lx5 file).

Reply:
Give me just a minute and I'll check.

Reply:
I've updated the l5x to have the couple of global tags in it that I was still using from before I tried to make everything program scoped for portability.

I recently got a new processor and haven't bothered to configure it for testing till now so I found a couple bugs and I see what you mean about things that will have to be changed to test with a different processor.  

These test fixes and l5x fixes should have came through with the latest change I just pushed out.

The test "TestMembersList" appears to be processor specific - it is looking up a known UDT ID in the PLC I had developed on, but that isn't fixed across processors (and probably even across downloads!) so I'll have to do some work to actually lookup the ID before listing it.  So don't worry if that one fails.

The test "TestList" seems to fail on v20 firmware.  I'll have to do some work there to figure out why.  Don't worry if that one fails for you either, probably.

Reply:
I believe I fixed all the issues with tests (except the testmemberslist test as expected). I wanted to ask though if the reconnection test has always been a little flaky though. It succeeds probably 95% of the time, but I feel like there are still some race condition potentials for different goroutines around the disconnect function.

Reply:
I just ran it a hundred times on the old code against the two PLCs I've got (L27@v32 and L61@v20viaENBT) and it passed.

```
# against the L27
go test -count=100 -timeout 30s -run TestReconnection github.com/danomagnum/gologix/tests
ok      github.com/danomagnum/gologix/tests     3.261s

# against the L61
go test -count=100 -timeout 30s -run TestReconnection github.com/danomagnum/gologix/tests
ok      github.com/danomagnum/gologix/tests     6.401s
```

I cloned your repo (marked 13 hours ago) and ran the same
```
go test -v -count=1 -timeout 30s -run TestReconnection github.com/danomagnum/gologix/tests
=== RUN   TestReconnection
2024/06/26 21:11:52 INFO Session connected controllerIp=192.168.2.241 sessionHandle=7077893
2024/06/26 21:11:52 INFO successfully opened connection controllerIp=192.168.2.241 ConnectionSize=4000 OTNetworkConnectionId=4287120971
2024/06/26 21:11:52 INFO starting disconnection controllerIp=192.168.2.241
panic: test timed out after 30s
running tests:
        TestReconnection (30s)

goroutine 19 [running]:
testing.(*M).startAlarm.func1()
        /usr/local/go/src/testing/testing.go:2366 +0x385
created by time.goFunc
        /usr/local/go/src/time/sleep.go:177 +0x2d

goroutine 1 [chan receive]:
testing.(*T).Run(0xc0000c0340, {0x671853?, 0x0?}, 0x681040)
        /usr/local/go/src/testing/testing.go:1750 +0x3ab
testing.runTests.func1(0xc0000c0340)
        /usr/local/go/src/testing/testing.go:2161 +0x37
testing.tRunner(0xc0000c0340, 0xc0000d1c70)
        /usr/local/go/src/testing/testing.go:1689 +0xfb
testing.runTests(0xc000012108, {0x7f5b00, 0x21, 0x21}, {0x1?, 0x4cc56e?, 0x7fa100?})
        /usr/local/go/src/testing/testing.go:2159 +0x445
testing.(*M).Run(0xc000098500)
        /usr/local/go/src/testing/testing.go:2027 +0x68b
main.main()
        _testmain.go:111 +0x16c

goroutine 6 [sync.Mutex.Lock]:
sync.runtime_SemacquireMutex(0x6bd268?, 0x48?, 0x410525?)
        /usr/local/go/src/runtime/sema.go:77 +0x25
sync.(*Mutex).lockSlow(0xc0000e0238)
        /usr/local/go/src/sync/mutex.go:171 +0x15d
sync.(*Mutex).Lock(...)
        /usr/local/go/src/sync/mutex.go:90
github.com/danomagnum/gologix.(*Client).send_recv_data(0xc0000e01c0, 0x6f, {0xc00002d7d0, 0x1, 0x1})
        /home/dan/programs/gologix_pr3/gologix/sendreceive.go:79 +0xa5
github.com/danomagnum/gologix.(*Client).Disconnect(0xc0000e01c0)
        /home/dan/programs/gologix_pr3/gologix/disconnect.go:62 +0x507
github.com/danomagnum/gologix.(*Client).send(0xc0000e01c0, 0x70, {0xc00002dbd8, 0x1, 0x2000000000000?})
        /home/dan/programs/gologix_pr3/gologix/sendreceive.go:65 +0x566
github.com/danomagnum/gologix.(*Client).send_recv_data(0xc0000e01c0, 0x70?, {0xc00002dbd8?, 0x1?, 0x1?})
        /home/dan/programs/gologix_pr3/gologix/sendreceive.go:81 +0x112
github.com/danomagnum/gologix.(*Client).Read_single(0xc0000e01c0, {0x6794d8, 0x22}, 0xc4, 0x5)
        /home/dan/programs/gologix_pr3/gologix/read.go:331 +0x390
github.com/danomagnum/gologix.readArray[...](0xc0000e01c0, {0x6794d8, 0x22}, 0x5)
        /home/dan/programs/gologix_pr3/gologix/read.go:410 +0xbd
github.com/danomagnum/gologix.(*Client).Read(0xc0000e01c0, {0x6794d8, 0x22}, {0x601e20, 0xc0000121f8})
        /home/dan/programs/gologix_pr3/gologix/read.go:156 +0x499
github.com/danomagnum/gologix/tests.TestReconnection(0xc0000c04e0)
        /home/dan/programs/gologix_pr3/gologix/tests/reconnection_test.go:43 +0x345
testing.tRunner(0xc0000c04e0, 0x681040)
        /usr/local/go/src/testing/testing.go:1689 +0xfb
created by testing.(*T).Run in goroutine 1
        /usr/local/go/src/testing/testing.go:1742 +0x390

goroutine 18 [sync.Mutex.Lock]:
sync.runtime_SemacquireMutex(0x6bd268?, 0x38?, 0x410525?)
        /usr/local/go/src/runtime/sema.go:77 +0x25
sync.(*Mutex).lockSlow(0xc0000e0238)
        /usr/local/go/src/sync/mutex.go:171 +0x15d
sync.(*Mutex).Lock(...)
        /usr/local/go/src/sync/mutex.go:90
github.com/danomagnum/gologix.(*Client).send_recv_data(0xc0000e01c0, 0x70, {0xc000062cd8, 0x1, 0x1})
        /home/dan/programs/gologix_pr3/gologix/sendreceive.go:79 +0xa5
github.com/danomagnum/gologix.(*Client).GetAttrList(0xc0000e01c0, 0xac, 0x1, {0xc000014310, 0x5, 0x673082?})
        /home/dan/programs/gologix_pr3/gologix/getattr.go:182 +0x490
github.com/danomagnum/gologix.(*Client).keepalive(0xc0000e01c0)
        /home/dan/programs/gologix_pr3/gologix/connect.go:134 +0x191
created by github.com/danomagnum/gologix.(*Client).Connect in goroutine 6
        /home/dan/programs/gologix_pr3/gologix/connect.go:99 +0xa87
FAIL    github.com/danomagnum/gologix/tests     30.026s
FAIL
```

I didn't check if this version is before or after your fixes.

Reply:
I appreciate you checking that, unfortunately I had some of my changes only locally that fixed the deadlocked goroutines issue that you showed.
With where the code is at right now, I can still sometimes fail that test, but it is because the second overall read that is labeled as should fail, may or may not fail the read.
This is because by default the way I put the client together, connect triggers the keepalive. Sometimes, the keepalive will end up trying the dead socket instead of the second read which will trigger the disconnect sequence. In that case the second read will actually be successful.

Overall though, that random failure on the test does not signify bad code in my opinion. How I am choosing to work around it is to set the keepalive to off before connecting the client so that way the second read can be guaranteed to catch the dead socket rather than just usually catching it.
With that modification to the test, I have successfully run the test 1000 times with no failure.

Reply:
I agree.  It makes sense to turn the keepalive off for that test.